### PR TITLE
Restore abuse-monitor

### PR DESCRIPTION
**Pull request by mia-pi-git**
Also adds support for using commands in a room without being in the room. Heavily restricted to specific punishments only for this plugin.

### CODE REVIEW

File: server/chat-plugins/abuse-monitor.ts [hunk 1]
```diff
@@ -0,0 +1,663 @@
...
+ * We use Google's Perspective API to classify messages.
+ * @see https://perspectiveapi.com/
+ * by Mia.
+ * @author mia-pi-git
+ */
+
+import {FS, Utils, Net, ProcessManager, Repl} from '../../lib';
+import {Config} from '../config-loader';
+import {toID} from '../../sim/dex-data';
+import {throttling} from '../../tools/set-import/importer';
```
Comment (Zarel) [inline server/chat-plugins/abuse-monitor.ts:14]: ugh pls no

Comment (Zarel) [reply server/chat-plugins/abuse-monitor.ts:14]: Please don't change `set-import` or introduce new dependencies in crazy ways. `tools` should be self-contained.

Comment (Zarel) [reply server/chat-plugins/abuse-monitor.ts:14]: The throttle in there can't be audited, anyway. Easy enough to just reimplement; maybe put an implementation in `Utils`.

Comment (mia-pi-git) [reply server/chat-plugins/abuse-monitor.ts]: Okay, done.

Comment (mia-pi-git) [reply server/chat-plugins/abuse-monitor.ts]: Moved it to utils and restored set-import.

File: server/chat-plugins/abuse-monitor.ts [hunk 2]
```diff
@@ -0,0 +1,719 @@
...
+	private counts = [] as number[];
+	private period: number;
+	max: number;
+	constructor(max: number, period: number) {
+		this.max = max;
+		this.period = period;
+	}
+	shouldRequest(depth = 0) {
+		const now = time();
+		if (this.second !== now) {
```
Comment (Zarel) [inline server/chat-plugins/abuse-monitor.ts:97]: The counter is wrong if multiple seconds pass between requests.

Comment (mia-pi-git) [reply server/chat-plugins/abuse-monitor.ts:97]: Wrong how so? should we just have an interval that clears it once a second?

Comment (mia-pi-git) [reply server/chat-plugins/abuse-monitor.ts:97]: Or, wait, no, I know what to do.

Comment (mia-pi-git) [reply server/chat-plugins/abuse-monitor.ts:97]: Fixed.

File: server/chat-plugins/abuse-monitor.ts [hunk 3]
```diff
@@ -0,0 +1,727 @@
...
+	shouldRequest() {
+		const now = time();
+		if (this.second !== now) {
+			const diff = now - this.second;
+			if (diff <= this.period) {
+				for (let i = 0; i < (diff - 1); i++) {
+					this.counts.push(0);
+				}
+			} else {
+				this.counts = [];
```
Comment (Zarel) [inline server/chat-plugins/abuse-monitor.ts:104]: This is wrong, it should be `Array(this.period).fill(0)`.

Comment (mia-pi-git) [reply server/chat-plugins/abuse-monitor.ts:104]: Yeah, that's what i was just thinking about - see DMs.

Comment (mia-pi-git) [reply server/chat-plugins/abuse-monitor.ts]: Fixed.

File: server/chat-plugins/abuse-monitor.ts [hunk 4]
```diff
@@ -0,0 +1,727 @@
...
+			} else {
+				this.counts = Array(this.period).fill(0);
+			}
+			this.counts.push(this.countThisSecond);
+			this.trimCounts();
+			this.second = now;
+			this.countThisSecond = 0;
+		}
+		const count = this.countThisSecond + 1;
+		const counts = this.getCounts();
```
Comment (Zarel) [inline server/chat-plugins/abuse-monitor.ts:112]: Keeping these separate isn't really worth the additional complexity.

File: server/chat-plugins/abuse-monitor.ts [hunk 5]
```diff
@@ -0,0 +1,727 @@
...
+		this.trimCounts();
+		return this.counts;
+	}
+	private trimCounts() {
+		if (this.counts.length > this.period) {
+			// oldest first
+			while (this.counts.length > this.period) this.counts.shift();
+		}
+	}
+}
```
Comment (Zarel) [inline server/chat-plugins/abuse-monitor.ts:140]:
```suggestion
export class RollingCounter {
	counts: number[] = [0];
	readonly size: number;
	constructor(limit: number) {
		this.size = limit;
	}
	increment() {
		this.counts[this.counts.length - 1]++;
	}
	rollOver(amount: number) {
		if (amount > this.size) {
			this.counts = Array(this.size).fill(0);
			return;
		}
		for (let i = 0; i < amount; i++) {
			this.counts.push(0);
			if (this.counts.length > this.size) this.counts.shift();
		}
	}
	mean() {
		let total = 0;
		for (const elem of this.counts) total++;
		return total / this.counts.length;
	}
}

export class Limiter {
	readonly counter: RollingCounter;
	readonly max: number;
	lastCounterRoll = time();
	constructor(max: number, period: number) {
		this.max = max;
		this.counter = new RollingCounter(period);
	}
	shouldRequest() {
		const now = time();
		this.counter.rollOver(now - this.lastCounterRoll);
		this.lastCounterRoll = now;

		if (this.counter.mean() > this.max) return false;
		this.counter.increment();
		return true;
	}
}
```

File: server/chat-plugins/abuse-monitor.ts [hunk 6]
```diff
@@ -0,0 +1,727 @@
...
+
+function isCommon(message: string) {
+	message = message.replace(/\?!\., ;:/ig, '');
```
Comment (Zarel) [inline server/chat-plugins/abuse-monitor.ts:143]:
```suggestion
	message = message.toLowerCase().replace(/\?!\., ;:/g, '');
```

Comment (mia-pi-git) [reply server/chat-plugins/abuse-monitor.ts]: Don't you want `/i`? What if someone types `WP`?

File: server/chat-plugins/abuse-monitor.ts (hunk 4 above)
Comment (mia-pi-git) [reply server/chat-plugins/abuse-monitor.ts]: Resolved by reimplementation.

File: server/chat-plugins/abuse-monitor.ts [hunk 7]
```diff
@@ -0,0 +1,715 @@
...
+			saveSettings();
+			this.privateGlobalModAction(`${this.user.name} set the abuse monitor trigger threshold to ${num}.`);
+			this.globalModlog('ABUSEMONITOR THRESHOLD', null, `${num}`);
+			this.sendReply(
+				`|html|Remember to use <code>/am respawn</code> to deploy the settings to the child process.`
+			);
+		},
+		resolve(target) {
+			this.checkCan('lock');
+			target = target.toLowerCase().trim().replace(/ /ig, '');
```
Comment (Zarel) [inline server/chat-plugins/abuse-monitor.ts:365]:
```suggestion
			target = target.toLowerCase().trim().replace(/ +/g, '');
```

File: server/chat-plugins/abuse-monitor.ts [hunk 8]
```diff
@@ -0,0 +1,715 @@
...
+	"INSULT": {},
+	"PROFANITY": {},
+	"THREAT": {},
+};
+
+export const cache: {
+	[roomid: string]: {[userid: string]: number} & {
+		staffNotified?: boolean,
+		claimed?: ID,
+	},
```
Comment (Zarel) [inline server/chat-plugins/abuse-monitor.ts:31]: Why is this `&` rather than a `users` sub-object?

Comment (mia-pi-git) [reply server/chat-plugins/abuse-monitor.ts:31]: I thought it was more readable, personally.

Comment (mia-pi-git) [reply server/chat-plugins/abuse-monitor.ts:31]: Nope, that's actually necessary for TS too.

Comment (mia-pi-git) [reply server/chat-plugins/abuse-monitor.ts]: Discussed in dms, restored cache[roomid].users.

File: server/chat-plugins/abuse-monitor.ts [hunk 9]
```diff
@@ -0,0 +1,726 @@
...
+					`Invalid punishment given. ` +
+					`Must be one of ${Object.keys(NOJOIN_COMMAND_WHITELIST).join(', ')}.`
+				);
+			}
+			this.room = tarRoom;
+			this.room.reportJoin('j', user.getIdentityWithStatus(this.room), user);
+			const result = await this.parse(`${cmd} ${rest}`, {bypassRoomCheck: true});
+			if (result) { // command succeeded - send followup
+				this.addModAction(
+					'If you have questions about this action, please contact staff by typing ``/ht`` in any chat.'
```
Comment (Zarel) [inline server/chat-plugins/abuse-monitor.ts:393]: Probably this should be a button.

**Review by Zarel — APPROVED** [review]

**Comment by Zarel** [comment]
Note that this adds support for using commands in rooms you're not in.

You should really not sneak in features like these. I approved it in DMs, but it should be clearly explained in the PR summary.

File: server/chat-plugins/abuse-monitor.ts (hunk 9 above)
Comment (mia-pi-git) [reply server/chat-plugins/abuse-monitor.ts]: Done.

**Pull request merged**